File: ameen-app/windows/Services/KdbxImportService.cs
using System.Diagnostics;
using System.IO.Compression;
using System.Security.Cryptography;
using System.Text;
using System.Xml.Linq;
using Ameen.Windows.Models;

namespace Ameen.Windows.Services;

public class KdbxImportResult
{
    public List<VaultItem> Items { get; set; } = new();
    public string DatabaseName { get; set; } = string.Empty;
    public int EntryCount { get; set; }
    public int GroupCount { get; set; }
}

public class KdbxImportService
{
    // KDBX signature constants
    private const uint KdbxSig1 = 0x9AA2D903;
    private const uint KdbxSig2 = 0xB54BFB67;
    private const uint KdbxVersion3_1 = 0x00030001;

    // Header field IDs
    private const byte HdrEnd = 0;
    private const byte HdrCipherId = 2;
    private const byte HdrCompression = 3;
    private const byte HdrMasterSeed = 4;
    private const byte HdrTransformSeed = 5;
    private const byte HdrTransformRounds = 6;
    private const byte HdrEncryptionIv = 7;
    private const byte HdrProtectedKey = 8;
    private const byte HdrStreamStart = 9;
    private const byte HdrInnerStreamId = 10;

    private static readonly byte[] CipherAes256 =
    {
        0x31, 0xC1, 0xF2, 0xE6, 0xBF, 0x71, 0x43, 0x50,
        0xBE, 0x58, 0x05, 0x21, 0x6A, 0xFC, 0x5A, 0xFF,
    };

    private const uint StreamSalsa20 = 2;

    /// <summary>
    /// Import entries from a KeePass KDBX3.1 database file.
    /// </summary>
    public async Task<KdbxImportResult> ImportAsync()
    {
        try
        {
            var result = await FilePicker.Default.PickAsync(new PickOptions
            {
                PickerTitle = "اختر ملف KeePass (.kdbx)",
                FileTypes = new FilePickerFileType(new Dictionary<DevicePlatform, IEnumerable<string>>
                {
                    { DevicePlatform.WinUI, new[] { ".kdbx" } },
                }),
            });

            if (result == null)
            {
                Debug.WriteLine("[KdbxImportService] File picker cancelled.");
                return new KdbxImportResult();
            }

            // Prompt for password (handled at UI level — here we accept bytes directly)
            // For now, the caller must provide password. We'll create a helper.
            Debug.WriteLine("[KdbxImportService] KDBX file selected — password must be provided by UI.");
            return new KdbxImportResult();
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"[KdbxImportService] Import failed: {ex.Message}");
            throw;
        }
    }

    /// <summary>
    /// Read KDBX3.1 file bytes with master password and extract vault items.
    /// </summary>
    public KdbxImportResult ReadKdbx(byte[] kdbxBytes, string password)
    {
        try
        {
            if (kdbxBytes.Length < 12)
                throw new InvalidOperationException("KDBX file is too small.");

            using var ms = new MemoryStream(kdbxBytes);
            using var reader = new BinaryReader(ms);

            uint sig1 = reader.ReadUInt32();
            uint sig2 = reader.ReadUInt32();
            uint version = reader.ReadUInt32();

            if (sig1 != KdbxSig1 || sig2 != KdbxSig2)
                throw new InvalidOperationException("Not a valid KDBX file.");

            if (version != KdbxVersion3_1)
                throw new InvalidOperationException($"Unsupported KDBX version: 0x{version:X8}. Only KDBX 3.1 is supported.");

            // Parse header
            var header = ParseHeader(reader);
            var encryptedData = ReadRemainingBytes(reader);

            // Derive key
            byte[] aesKey = DeriveKey(password, header.TransformSeed, header.TransformRounds, header.MasterSeed);

            // Decrypt
            byte[] decrypted = DecryptAesCbc(aesKey, header.EncryptionIv, encryptedData);

            // Verify stream start bytes
            if (header.StreamStartBytes != null && header.StreamStartBytes.Length > 0)
            {
                for (int i = 0; i < Math.Min(header.StreamStartBytes.Length, decrypted.Length); i++)
                {
                    if (decrypted[i] != header.StreamStartBytes[i])
                        throw new InvalidOperationException("Wrong password — stream start bytes mismatch.");
                }
            }

            // Decompress
            byte[] xmlBytes;
            if (header.Compression == 1)
            {
                xmlBytes = GzipDecompress(decrypted);
            }
            else
            {
                xmlBytes = decrypted;
            }

            string xmlText = Encoding.UTF8.GetString(xmlBytes);
            var entries = ParseXml(xmlText, header.ProtectedStreamKey);

            var result = new KdbxImportResult
            {
                Items = entries.Select((e, i) => new VaultItem
                {
                    Id = Guid.NewGuid().ToString("N")[..12],
                    Type = "password",
                    Title = !string.IsNullOrEmpty(e.title) ? e.title : (!string.IsNullOrEmpty(e.url) ? e.url : $"مستورد {i + 1}"),
                    Username = e.username,
                    Password = e.password,
                    Url = e.url,
                    Notes = e.notes,
                    UpdatedAt = DateTime.Now,
                }).ToList(),
                DatabaseName = ExtractDatabaseName(xmlText),
                EntryCount = entries.Count,
                GroupCount = CountGroups(xmlText),
            };

            Debug.WriteLine($"[KdbxImportService] Imported {result.EntryCount} entries from '{result.DatabaseName}'.");
            return result;
        }
        catch (CryptographicException)
        {
            throw new InvalidOperationException("فشل فك تشفير ملف KeePass — كلمة المرور غير صحيحة.");
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"[KdbxImportService] Read error: {ex.Message}");
            throw;
        }
    }

    // ============================================================
    // Header Parsing
    // ============================================================

    private class KdbxHeader
    {
        public byte[]? MasterSeed;
        public byte[]? TransformSeed;
        public ulong TransformRounds;
        public byte[]? EncryptionIv;
        public byte[]? ProtectedStreamKey;
        public byte[]? StreamStartBytes;
        public uint InnerStreamId;
        public uint Compression;
    }

    private static KdbxHeader ParseHeader(BinaryReader reader)
    {
        var header = new KdbxHeader();

        while (true)
        {
            byte fieldType = reader.ReadByte();
            ushort fieldSize = reader.ReadUInt16();

            byte[] fieldData;
            if (fieldSize > 0)
            {
                fieldData = reader.ReadBytes(fieldSize);
            }
            else
            {
                fieldData = Array.Empty<byte>();
            }

            switch (fieldType)
            {
                case HdrEnd:
                    return header;

                case HdrCipherId:
                    if (fieldData.Length != 16)
                        throw new InvalidOperationException("Invalid cipher ID.");
                    for (int i = 0; i < 16; i++)
                    {
                        if (fieldData[i] != CipherAes256[i])
                            throw new InvalidOperationException($"Unsupported cipher. Only AES-256 is supported.");
                    }
                    break;

                case HdrCompression:
                    if (fieldSize >= 4)
                        header.Compression = BitConverter.ToUInt32(fieldData, 0);
                    break;

                case HdrMasterSeed:
                    header.MasterSeed = fieldData;
                    break;

                case HdrTransformSeed:
                    header.TransformSeed = fieldData;
                    break;

                case HdrTransformRounds:
                    header.TransformRounds = BitConverter.ToUInt64(fieldData, 0);
                    break;

                case HdrEncryptionIv:
                    header.EncryptionIv = fieldData;
                    break;

                case HdrProtectedKey:
                    header.ProtectedStreamKey = fieldData;
                    break;

                case HdrStreamStart:
                    header.StreamStartBytes = fieldData;
                    break;

                case HdrInnerStreamId:
                    if (fieldSize >= 4)
                        header.InnerStreamId = BitConverter.ToUInt32(fieldData, 0);
                    break;
            }
        }
    }

    private static byte[] ReadRemainingBytes(BinaryReader reader)
    {
        using var ms = new MemoryStream();
        reader.BaseStream.CopyTo(ms);
        return ms.ToArray();
    }

    // ============================================================
    // Key Derivation (KeePass AES-ECB transform)
    // ============================================================

    private static byte[] DeriveKey(string password, byte[] transformSeed, ulong transformRounds, byte[] masterSeed)
    {
        byte[] compositeKey = SHA256.HashData(Encoding.UTF8.GetBytes(password));

        using var aes = Aes.Create();
        aes.Key = compositeKey;
        aes.Mode = CipherMode.ECB;
        aes.Padding = PaddingMode.None;

        byte[] transformed = new byte[transformSeed.Length];
        Array.Copy(transformSeed, transformed, transformSeed.Length);

        using var encryptor = aes.CreateEncryptor();
        for (ulong round = 0; round < transformRounds; round++)
        {
            transformed = encryptor.TransformFinalBlock(transformed, 0, transformed.Length);
        }

        byte[] combined = new byte[transformed.Length + masterSeed.Length];
        Array.Copy(transformed, 0, combined, 0, transformed.Length);
        Array.Copy(masterSeed, 0, combined, transformed.Length, masterSeed.Length);

        return SHA256.HashData(combined);
    }

    // ============================================================
    // AES-256-CBC
    // ============================================================

    private static byte[] DecryptAesCbc(byte[] key, byte[] iv, byte[] data)
    {
        using var aes = Aes.Create();
        aes.Key = key;
        aes.Mode = CipherMode.CBC;
        aes.Padding = PaddingMode.None;
        aes.IV = iv;

        using var decryptor = aes.CreateDecryptor();
        return decryptor.TransformFinalBlock(data, 0, data.Length);
    }

    // ============================================================
    // GZip Decompression
    // ============================================================

    private static byte[] GzipDecompress(byte[] data)
    {
        using var input = new MemoryStream(data);
        using var gzip = new GZipStream(input, CompressionMode.Decompress);
        using var output = new MemoryStream();
        gzip.CopyTo(output);
        return output.ToArray();
    }

    // ============================================================
    // XML Parsing
    // ============================================================

    private class KdbxEntry
    {
        public string title = string.Empty;
        public string username = string.Empty;
        public string password = string.Empty;
        public string url = string.Empty;
        public string notes = string.Empty;
    }

    private static List<KdbxEntry> ParseXml(string xml, byte[]? protectedKey)
    {
        var entries = new List<KdbxEntry>();
        var doc = XDocument.Parse(xml);
        var ns = doc.Root!.GetDefaultNamespace();

        foreach (var entryElem in doc.Descendants(ns + "Entry"))
        {
            var entry = new KdbxEntry();

            foreach (var stringElem in entryElem.Descendants(ns + "String"))
            {
                var keyElem = stringElem.Element(ns + "Key");
                var valueElem = stringElem.Element(ns + "Value");

                if (keyElem == null || valueElem == null) continue;

                string? key = keyElem.Value;
                string? value = valueElem.Value;

                // Check if protected
                var protectedAttr = valueElem.Attribute("Protected");
                bool isProtected = protectedAttr != null && protectedAttr.Value == "True";

                string decrypted;
                if (isProtected && protectedKey != null && protectedKey.Length > 0)
                {
                    decrypted = DecryptProtectedString(protectedKey, value ?? "");
                }
                else
                {
                    decrypted = value ?? "";
                }

                switch (key)
                {
                    case "Title": entry.title = decrypted; break;
                    case "UserName": entry.username = decrypted; break;
                    case "Password": entry.password = decrypted; break;
                    case "URL": entry.url = decrypted; break;
                    case "Notes": entry.notes = decrypted; break;
                }
            }

            if (!string.IsNullOrEmpty(entry.title) ||
                !string.IsNullOrEmpty(entry.username) ||
                !string.IsNullOrEmpty(entry.password))
            {
                entries.Add(entry);
            }
        }

        return entries;
    }

    // ============================================================
    // Salsa20 Decryption
    // ============================================================

    private static string DecryptProtectedString(byte[] key, string base64Value)
    {
        try
        {
            byte[] raw = Convert.FromBase64String(base64Value);
            if (raw.Length < 8) return base64Value;

            byte[] iv = raw.Take(8).ToArray();
            byte[] ciphertext = raw.Skip(8).ToArray();

            byte[] plaintext = Salsa20Decrypt(key, iv, ciphertext);

            // Remove null padding
            int end = plaintext.Length;
            while (end > 0 && plaintext[end - 1] == 0) end--;

            return Encoding.UTF8.GetString(plaintext, 0, end);
        }
        catch
        {
            return base64Value;
        }
    }

    private static byte[] Salsa20Decrypt(byte[] key, byte[] iv, byte[] data)
    {
        byte[] output = new byte[data.Length];
        byte[] block = new byte[64];
        int pos = 0;
        ulong counter = 0;

        while (pos < data.Length)
        {
            GenerateSalsa20Block(key, iv, counter, block);
            counter++;

            int remaining = Math.Min(64, data.Length - pos);
            for (int i = 0; i < remaining; i++)
            {
                output[pos + i] = (byte)(data[pos + i] ^ block[i]);
            }
            pos += 64;
        }

        return output;
    }

    private static void GenerateSalsa20Block(byte[] key, byte[] iv, ulong counter, byte[] output)
    {
        uint[] state = new uint[16];

        // "expand 32-byte k"
        state[0] = 0x61707865;
        state[5] = 0x3320646E;
        state[10] = 0x79622D32;
        state[15] = 0x6B206574;

        // Key (8 x uint32)
        state[1] = BitConverter.ToUInt32(key, 0);
        state[2] = BitConverter.ToUInt32(key, 4);
        state[3] = BitConverter.ToUInt32(key, 8);
        state[4] = BitConverter.ToUInt32(key, 12);

        state[11] = BitConverter.ToUInt32(key, 16);
        state[12] = BitConverter.ToUInt32(key, 20);
        state[13] = BitConverter.ToUInt32(key, 24);
        state[14] = BitConverter.ToUInt32(key, 28);

        // IV (2 x uint32)
        state[6] = BitConverter.ToUInt32(iv, 0);
        state[7] = BitConverter.ToUInt32(iv, 4);

        // Counter
        state[8] = (uint)(counter & 0xFFFFFFFF);
        state[9] = (uint)(counter >> 32);

        uint[] working = (uint[])state.Clone();

        // 20 rounds (10 double-rounds)
        for (int i = 0; i < 10; i++)
        {
            // Column rounds
            QuarterRound(ref working, 0, 4, 8, 12);
            QuarterRound(ref working, 5, 9, 13, 1);
            QuarterRound(ref working, 10, 14, 2, 6);
            QuarterRound(ref working, 15, 3, 7, 11);

            // Row rounds
            QuarterRound(ref working, 0, 1, 2, 3);
            QuarterRound(ref working, 5, 6, 7, 4);
            QuarterRound(ref working, 10, 11, 8, 9);
            QuarterRound(ref working, 15, 12, 13, 14);
        }

        // Add original state
        for (int i = 0; i < 16; i++)
        {
            working[i] += state[i];
        }

        // Write to bytes
        for (int i = 0; i < 16; i++)
        {
            byte[] bytes = BitConverter.GetBytes(working[i]);
            Array.Copy(bytes, 0, output, i * 4, 4);
        }
    }

    private static void QuarterRound(ref uint[] state, int a, int b, int c, int d)
    {
        state[b] ^= RotateLeft(state[a] + state[d], 7);
        state[c] ^= RotateLeft(state[b] + state[a], 9);
        state[d] ^= RotateLeft(state[c] + state[b], 13);
        state[a] ^= RotateLeft(state[d] + state[c], 18);
    }

    private static uint RotateLeft(uint value, int shift)
    {
        return (value << shift) | (value >> (32 - shift));
    }

    private static string ExtractDatabaseName(string xml)
    {
        var doc = XDocument.Parse(xml);
        var ns = doc.Root!.GetDefaultNamespace();
        var nameElem = doc.Descendants(ns + "DatabaseName").FirstOrDefault();
        return nameElem?.Value.Trim() ?? "KeePass";
    }

    private static int CountGroups(string xml)
    {
        var doc = XDocument.Parse(xml);
        var ns = doc.Root!.GetDefaultNamespace();
        return doc.Descendants(ns + "Group").Count();
    }
}

File: ameen-app/android/app/src/main/java/com/ameen/app/services/KdbxImportService.kt
package com.ameen.app.services

import android.content.Context
import android.net.Uri
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.security.MessageDigest
import java.util.UUID
import java.util.zip.GZIPInputStream
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

class KdbxImportService(private val context: Context) {

    companion object {
        private const val TAG = "KdbxImportService"
        private const val KDBX_SIG1 = 0x9AA2D903L
        private const val KDBX_SIG2 = 0xB54BFB67L
        private const val KDBX_VER_3_1 = 0x00030001L

        private const val HDR_END = 0
        private const val HDR_CIPHER_ID = 2
        private const val HDR_COMPRESSION = 3
        private const val HDR_MASTER_SEED = 4
        private const val HDR_TRANSFORM_SEED = 5
        private const val HDR_TRANSFORM_ROUNDS = 6
        private const val HDR_ENCRYPTION_IV = 7
        private const val HDR_PROTECTED_KEY = 8
        private const val HDR_STREAM_START = 9
        private const val HDR_INNER_STREAM_ID = 10

        private val CIPHER_AES256 = byteArrayOf(
            0x31.toByte(), 0xC1.toByte(), 0xF2.toByte(), 0xE6.toByte(),
            0xBF.toByte(), 0x71.toByte(), 0x43.toByte(), 0x50.toByte(),
            0xBE.toByte(), 0x58.toByte(), 0x05.toByte(), 0x21.toByte(),
            0x6A.toByte(), 0xFC.toByte(), 0x5A.toByte(), 0xFF.toByte()
        )
    }

    data class KdbxImportResult(
        val items: JSONArray = JSONArray(),
        val databaseName: String = "",
        val entryCount: Int = 0
    )

    suspend fun importFromUri(uri: Uri, password: String): KdbxImportResult = withContext(Dispatchers.IO) {
        try {
            val bytes = context.contentResolver.openInputStream(uri)?.use { it.readBytes() }
                ?: throw IllegalStateException("تعذر قراءة ملف KDBX")
            readKdbx(bytes, password)
        } catch (e: Exception) {
            Log.e(TAG, "فشل استيراد KDBX: ${e.message}", e)
            throw e
        }
    }

    fun readKdbx(kdbxBytes: ByteArray, password: String): KdbxImportResult {
        try {
            if (kdbxBytes.size < 12) throw IllegalStateException("ملف KDBX صغير جداً")

            val buf = kdbxBytes.inputStream().buffered()
            val sig1 = readUInt32(buf)
            val sig2 = readUInt32(buf)
            val ver = readUInt32(buf)

            if (sig1 != KDBX_SIG1 || sig2 != KDBX_SIG2)
                throw IllegalStateException("ليس ملف KDBX صالح")
            if (ver != KDBX_VER_3_1)
                throw IllegalStateException("إصدار KDBX غير مدعوم — يدعم فقط 3.1")

            val header = parseHeader(buf)
            val encryptedPayload = buf.readBytes()

            val aesKey = deriveKey(password, header.transformSeed!!, header.transformRounds, header.masterSeed!!)

            val decrypted = decryptAesCbc(aesKey, header.encryptionIv!!, encryptedPayload)

            header.streamStartBytes?.let { start ->
                for (i in start.indices) {
                    if (i < decrypted.size && decrypted[i] != start[i])
                        throw IllegalStateException("كلمة المرور غير صحيحة")
                }
            }

            val xmlBytes = if (header.compression == 1L) gzipDecompress(decrypted) else decrypted
            val xmlText = String(xmlBytes, Charsets.UTF_8)

            return parseXml(xmlText, header.protectedStreamKey ?: ByteArray(32))
        } catch (e: Exception) {
            Log.e(TAG, "فشل قراءة KDBX: ${e.message}", e)
            throw e
        }
    }

    // ============================================================
    // Header Parsing
    // ============================================================

    data class KdbxHeader(
        var masterSeed: ByteArray? = null,
        var transformSeed: ByteArray? = null,
        var transformRounds: Long = 0,
        var encryptionIv: ByteArray? = null,
        var protectedStreamKey: ByteArray? = null,
        var streamStartBytes: ByteArray? = null,
        var compression: Long = 0
    )

    private fun parseHeader(buf: java.io.InputStream): KdbxHeader {
        val header = KdbxHeader()

        while (true) {
            val fieldType = buf.read()
            if (fieldType < 0) break
            if (fieldType == HDR_END) break

            val fieldSize = (buf.read() or (buf.read() shl 8))
            val fieldData = ByteArray(fieldSize)
            buf.read(fieldData)

            when (fieldType) {
                HDR_CIPHER_ID -> {
                    if (!fieldData.contentEquals(CIPHER_AES256))
                        throw IllegalStateException("نوع تشفير غير مدعوم — يتطلب AES-256")
                }
                HDR_COMPRESSION -> {
                    if (fieldSize >= 4)
                        header.compression = (fieldData[0].toLong() and 0xFF) or
                            ((fieldData[1].toLong() and 0xFF) shl 8) or
                            ((fieldData[2].toLong() and 0xFF) shl 16) or
                            ((fieldData[3].toLong() and 0xFF) shl 24)
                }
                HDR_MASTER_SEED -> header.masterSeed = fieldData
                HDR_TRANSFORM_SEED -> header.transformSeed = fieldData
                HDR_TRANSFORM_ROUNDS -> {
                    var rounds = 0L
                    for (i in 7 downTo 0) {
                        val b = if (i < fieldSize) fieldData[i].toLong() and 0xFF else 0
                        rounds = (rounds shl 8) or b
                    }
                    header.transformRounds = rounds
                }
                HDR_ENCRYPTION_IV -> header.encryptionIv = fieldData
                HDR_PROTECTED_KEY -> header.protectedStreamKey = fieldData
                HDR_STREAM_START -> header.streamStartBytes = fieldData
            }
        }
        return header
    }

    // ============================================================
    // Key Derivation
    // ============================================================

    private fun deriveKey(
        password: String,
        transformSeed: ByteArray,
        transformRounds: Long,
        masterSeed: ByteArray
    ): ByteArray {
        val compositeKey = MessageDigest.getInstance("SHA-256")
            .digest(password.toByteArray(Charsets.UTF_8))

        val cipher = Cipher.getInstance("AES/ECB/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(compositeKey, "AES"))

        var transformed = transformSeed.clone()
        for (round in 0 until transformRounds) {
            transformed = cipher.doFinal(transformed)
        }

        val combined = transformed + masterSeed
        return MessageDigest.getInstance("SHA-256").digest(combined)
    }

    // ============================================================
    // AES-256-CBC
    // ============================================================

    private fun decryptAesCbc(key: ByteArray, iv: ByteArray, data: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("AES/CBC/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), IvParameterSpec(iv))
        return cipher.doFinal(data)
    }

    // ============================================================
    // GZip
    // ============================================================

    private fun gzipDecompress(data: ByteArray): ByteArray {
        GZIPInputStream(ByteArrayInputStream(data)).use { gz ->
            ByteArrayOutputStream().use { out ->
                gz.copyTo(out)
                return out.toByteArray()
            }
        }
    }

    // ============================================================
    // Salsa20
    // ============================================================

    private fun salsa20Decrypt(key: ByteArray, iv: ByteArray, data: ByteArray): ByteArray {
        val output = ByteArray(data.size)
        val block = ByteArray(64)
        var counter = 0L

        var pos = 0
        while (pos < data.size) {
            generateSalsa20Block(key, iv, counter, block)
            counter++
            val remaining = minOf(64, data.size - pos)
            for (i in 0 until remaining) {
                output[pos + i] = (data[pos + i].toInt() xor block[i].toInt()).toByte()
            }
            pos += 64
        }
        return output
    }

    private fun generateSalsa20Block(key: ByteArray, iv: ByteArray, counter: Long, output: ByteArray) {
        val state = IntArray(16)
        state[0] = 0x61707865
        state[5] = 0x3320646E
        state[10] = 0x79622D32
        state[15] = 0x6B206574

        for (i in 0..3) state[1 + i] = bytesToInt(key, i * 4)
        state[6] = bytesToInt(iv, 0)
        state[7] = bytesToInt(iv, 4)
        state[8] = (counter and 0xFFFFFFFFL).toInt()
        state[9] = ((counter ushr 32) and 0xFFFFFFFFL).toInt()
        for (i in 0..3) state[11 + i] = bytesToInt(key, 16 + i * 4)

        val working = state.clone()

        for (i in 0 until 10) {
            qr(working, 0, 4, 8, 12)
            qr(working, 5, 9, 13, 1)
            qr(working, 10, 14, 2, 6)
            qr(working, 15, 3, 7, 11)
            qr(working, 0, 1, 2, 3)
            qr(working, 5, 6, 7, 4)
            qr(working, 10, 11, 8, 9)
            qr(working, 15, 12, 13, 14)
        }

        for (i in 0 until 16) {
            val v = (working[i] + state[i])
            intToBytes(v, output, i * 4)
        }
    }

    private fun qr(state: IntArray, a: Int, b: Int, c: Int, d: Int) {
        state[b] = state[b] xor ((state[a] + state[d]) shl 7 or ((state[a] + state[d]) ushr 25))
        state[c] = state[c] xor ((state[b] + state[a]) shl 9 or ((state[b] + state[a]) ushr 23))
        state[d] = state[d] xor ((state[c] + state[b]) shl 13 or ((state[c] + state[b]) ushr 19))
        state[a] = state[a] xor ((state[d] + state[c]) shl 18 or ((state[d] + state[c]) ushr 14))
    }

    private fun bytesToInt(bytes: ByteArray, offset: Int): Int {
        return ((bytes[offset].toInt() and 0xFF))
            or ((bytes[offset + 1].toInt() and 0xFF) shl 8)
            or ((bytes[offset + 2].toInt() and 0xFF) shl 16)
            or ((bytes[offset + 3].toInt() and 0xFF) shl 24)
    }

    private fun intToBytes(value: Int, out: ByteArray, offset: Int) {
        out[offset] = (value and 0xFF).toByte()
        out[offset + 1] = ((value ushr 8) and 0xFF).toByte()
        out[offset + 2] = ((value ushr 16) and 0xFF).toByte()
        out[offset + 3] = ((value ushr 24) and 0xFF).toByte()
    }

    // ============================================================
    // XML Parsing
    // ============================================================

    private fun parseXml(xml: String, protectedKey: ByteArray): KdbxImportResult {
        val items = JSONArray()
        val entryRegex = Regex("<Entry>([\\s\\S]*?)</Entry>")
        val dbNameMatch = Regex("<DatabaseName>(.*?)</DatabaseName>").find(xml)
        val dbName = dbNameMatch?.groupValues?.get(1)?.trim() ?: "KeePass"

        for (match in entryRegex.findAll(xml)) {
            val entryBlock = match.groupValues[1]
            val title = extractString(entryBlock, "Title", protectedKey)
            val username = extractString(entryBlock, "UserName", protectedKey)
            val password = extractString(entryBlock, "Password", protectedKey)
            val url = extractString(entryBlock, "URL", protectedKey)
            val notes = extractString(entryBlock, "Notes", protectedKey)

            if (title.isNotEmpty() || username.isNotEmpty() || password.isNotEmpty()) {
                items.put(JSONObject().apply {
                    put("id", UUID.randomUUID().toString().take(12))
                    put("type", "password")
                    put("title", title.ifEmpty { url.ifEmpty { "مستورد" } })
                    put("username", username)
                    put("password", password)
                    put("url", url)
                    put("content", "")
                    put("cardholder", "")
                    put("number", "")
                    put("expiry", "")
                    put("cvv", "")
                    put("notes", notes)
                    put("otpSecret", "")
                    put("updatedAt", System.currentTimeMillis())
                    put("isConflict", false)
                })
            }
        }

        Log.d(TAG, "تم استيراد ${items.length()} عنصر من '$dbName'")
        return KdbxImportResult(items, dbName, items.length())
    }

    private fun extractString(entryBlock: String, key: String, protectedKey: ByteArray): String {
        val stringRegex = Regex("<String>[\\s\\S]*?<Key>$key</Key>[\\s\\S]*?<Value([^>]*)>([\\s\\S]*?)</Value>[\\s\\S]*?</String>")
        val match = stringRegex.find(entryBlock) ?: return ""

        val attrs = match.groupValues[1]
        val value = match.groupValues[2]
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")
            .replace("&apos;", "'")

        val isProtected = attrs.contains("Protected=\"True\"", ignoreCase = true)
        return if (isProtected) decryptProtectedValue(protectedKey, value) else value
    }

    private fun decryptProtectedValue(protectedKey: ByteArray, base64Value: String): String {
        return try {
            val raw = android.util.Base64.decode(base64Value, android.util.Base64.DEFAULT)
            if (raw.size < 8) return base64Value
            val iv = raw.copyOfRange(0, 8)
            val ciphertext = raw.copyOfRange(8, raw.size)
            val plaintext = salsa20Decrypt(protectedKey, iv, ciphertext)
            val end = plaintext.indexOfLast { it != 0.toByte() } + 1
            String(plaintext, 0, end.coerceAtLeast(0), Charsets.UTF_8)
        } catch (e: Exception) {
            base64Value
        }
    }

    private fun readUInt32(buf: java.io.InputStream): Long {
        val b1 = buf.read().toLong()
        val b2 = buf.read().toLong()
        val b3 = buf.read().toLong()
        val b4 = buf.read().toLong()
        return (b1 and 0xFF) or ((b2 and 0xFF) shl 8) or ((b3 and 0xFF) shl 16) or ((b4 and 0xFF) shl 24)
    }
}
